### PR TITLE
[FIX] Propagate all *_cls/*_class plugin metadata fields in PluginManager

### PR DESCRIPTION
**Pull request by jaseemjaskp**
## What

- `PluginManager.load_plugins()` (in `unstract/core`) only copied a hardcoded 3-field allowlist (`entrypoint_cls`, `exception_cls`, `service_class`) from a plugin's `metadata` dict into the registered plugin dict.
- Generalized this to copy any metadata key ending in `_cls` or `_class`, so new plugin fields don't require this shared loader to be updated every time.

## Why

- The `verticals_usage` plugin (enterprise/cloud-only, feeds API deployment usage into the API Hub subscription dashboard) defines a `headers_cache_class` metadata field that the loader's hardcoded allowlist didn't know about. It was silently dropped, so every `plugin["headers_cache_class"]` lookup in `backend/plugins/workflow_manager/workflow_v2/api_hub_usage_utils.py` raised `KeyError` — swallowed by broad `except Exception` blocks.
- Net effect: API Hub request headers were never cached in Redis, `store_usage()` (the only path that writes to the `verticals.subscription_usage` table) was never reached, and the API Hub portal's subscription dashboard stopped showing usage — with no visible errors anywhere.
- Traced the regression to PR #1736 ("refactor: Add dynamic plugin loading for enterprise components"), which switched `api_hub_usage_utils.py` from a build-time file-replacement model to a runtime `get_plugin()` dict lookup, without the shared `PluginManager` (added in #1618) being updated to support the `headers_cache_class` field.

## How

- Replaced the three explicit `if "x" in metadata: plugin_data["x"] = metadata["x"]` checks in `unstract/core/src/unstract/core/plugins/plugin_manager.py` with a loop that copies any metadata key matching `*_cls`/`*_class`.
- Added `unstract/core/tests/test_plugin_manager.py` covering: unknown `_class`/`_cls` fields (e.g. `headers_cache_class`) now propagate, existing `entrypoint_cls`/`exception_cls` still work, non-class metadata isn't duplicated at the top level, and disabled plugins are still skipped.

## Can this PR break any existing features. If yes, please list possible items. If no, please explain why. (PS: Admins do not merge the PR without this section filled)

- No. Verified via grep across `backend/` that no other plugin's metadata relies on a field outside the previous three-field allowlist (only `service_class` is used elsewhere — notifications, hubspot, subscription_usage, file_converter, payload_modifier, etc.), so existing plugin behavior is unchanged; this purely adds support for previously-dropped fields.

## Database Migrations

- None

## Env Config

- None

## Relevant Docs

- None

## Related Issues or PRs

- Regression introduced in #1736

## Dependencies Versions

- None

## Notes on Testing

- Added `unstract/core/tests/test_plugin_manager.py` (4 new tests, all passing).
- Ran full `unstract/core` test suite (`uv run pytest unstract/core/tests/`) — no new regressions; 2 pre-existing unrelated failures were already broken before this change.
- Ran `ruff check` / `ruff format --check` on changed files — clean.

## Checklist

I have read and understood the [Contribution Guidelines](https://docs.unstract.com/unstract/contributing/unstract/).

### CODE REVIEW

File: unstract/core/src/unstract/core/plugins/plugin_manager.py [hunk 1]
```diff
@@ -215,13 +215,13 @@ def load_plugins(self) -> None:
...
-                    plugin_data["entrypoint_cls"] = metadata["entrypoint_cls"]
-                if "exception_cls" in metadata:
-                    plugin_data["exception_cls"] = metadata["exception_cls"]
-                if "service_class" in metadata:
-                    plugin_data["service_class"] = metadata["service_class"]
+                # Add optional class/entrypoint fields if present (any key ending in
+                # _cls or _class, e.g. entrypoint_cls, exception_cls, service_class,
+                # headers_cache_class) so new plugin metadata fields don't need this
+                # loader updated every time.
+                for key, value in metadata.items():
```
Comment (jaseemjaskp) [inline unstract/core/src/unstract/core/plugins/plugin_manager.py:225]: **[Silent failure — Medium]** This loop propagates class handles *by name only* — it copies whatever keys end in `_cls`/`_class`, with no notion of a *required* field. That fixes the `headers_cache_class` drop, but the mechanism that made it invisible is untouched: if a plugin declares a misnamed field (`header_cache_class`, `headers_cache_clss`), the correct key is simply absent and `verticals_usage_plugin["headers_cache_class"]()` in `backend/plugins/workflow_manager/workflow_v2/api_hub_usage_utils.py:38,99` raises `KeyError` again — swallowed by the same broad `except Exception` (lines 51/74/101) this PR was written to defeat.

Consider having consumers read with `.get()` and log an explicit failure, or validating a plugin's required class fields at load time, so a typo errors loudly at startup instead of failing silently per-request. (Narrowing those `except Exception` blocks to distinguish a permanent misconfiguration from a transient runtime error is the more complete fix, but is outside this PR's changed lines.)

File: unstract/core/src/unstract/core/plugins/plugin_manager.py [hunk 2]
```diff
@@ -215,13 +215,13 @@ def load_plugins(self) -> None:
...
                 }
 
-                # Add optional fields if present
-                if "entrypoint_cls" in metadata:
-                    plugin_data["entrypoint_cls"] = metadata["entrypoint_cls"]
-                if "exception_cls" in metadata:
-                    plugin_data["exception_cls"] = metadata["exception_cls"]
-                if "service_class" in metadata:
-                    plugin_data["service_class"] = metadata["service_class"]
+                # Add optional class/entrypoint fields if present (any key ending in
```
Comment (jaseemjaskp) [inline unstract/core/src/unstract/core/plugins/plugin_manager.py]: **[Maintainability — Low]** The `_cls`/`_class` suffix is now load-bearing behavior but is documented only in this comment. Consider promoting it to a module-level constant (e.g. `_CLASS_HANDLE_SUFFIXES = ("_cls", "_class")`) so the contract is greppable and testable. Also consider dropping the `headers_cache_class` example from the comment — it names a single enterprise plugin's field, so it is the part most likely to rot; the commit message already records the motivating case.

File: unstract/core/src/unstract/core/plugins/plugin_manager.py [hunk 3]
```diff
@@ -215,13 +215,13 @@ def load_plugins(self) -> None:
...
+                # _cls or _class, e.g. entrypoint_cls, exception_cls, service_class,
+                # headers_cache_class) so new plugin metadata fields don't need this
+                # loader updated every time.
+                for key, value in metadata.items():
+                    if key.endswith(("_cls", "_class")) and key not in plugin_data:
```
Comment (jaseemjaskp) [inline unstract/core/src/unstract/core/plugins/plugin_manager.py]: **[Nit — Low]** `key not in plugin_data` can never be false here: at this point `plugin_data` holds only `version`/`module`/`metadata` (none end in `_cls`/`_class`) and metadata keys are already unique, so the guard is effectively dead. Fine to keep as future-proofing against a reserved key that someday ends in `_class`, but if so, add a one-line note saying it's purely defensive — otherwise it reads as load-bearing when it isn't. Flagged independently by the code-review, simplifier, and test passes.

File: unstract/core/tests/test_plugin_manager.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    _write_plugin(
+        plugins_pkg_dir,
+        "sample_plugin",
+        '{"name": "sample_plugin", "service_class": str, "headers_cache_class": dict, }',
+    )
+
+    plugin = _load(plugins_pkg_dir).get_plugin("sample_plugin")
+
+    assert plugin["service_class"] is str
+    assert plugin["headers_cache_class"] is dict
```
Comment (jaseemjaskp) [inline unstract/core/tests/test_plugin_manager.py:56]: **[Test gap — High value]** The real behavioral widening in this change is that the loop propagates a `*_cls`/`*_class` key *regardless of its value*, whereas the old code copied 3 known class fields. No test pins that contract. A future "only propagate real classes" `isinstance`/callable check would pass all 4 tests here while silently reverting the fix for a field whose value is `None`/lazy at load time.

Suggest adding a test with metadata `{"name": "sample_plugin", "headers_cache_class": None}` asserting `plugin["headers_cache_class"] is None` — this locks "propagate by name, not by value."

File: unstract/core/tests/test_plugin_manager.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+def test_disabled_plugin_is_not_loaded(plugins_pkg_dir):
+    _write_plugin(
+        plugins_pkg_dir,
+        "sample_plugin",
+        '{"name": "sample_plugin", "disable": True}',
+    )
+
+    manager = _load(plugins_pkg_dir)
+
+    assert not manager.has_plugin("sample_plugin")
```
Comment (jaseemjaskp) [inline unstract/core/tests/test_plugin_manager.py:95]: **[Test gap — Medium]** The skip condition at `plugin_manager.py:202` is `metadata.get("disable", False) or not metadata.get("is_active", True)`. This test exercises only the `disable=True` operand; the `is_active=False` branch is separate and could regress independently (e.g. someone flips the default or drops the operand). Suggest mirroring this test with metadata `{"name": "sample_plugin", "is_active": False}` and asserting `not manager.has_plugin("sample_plugin")`.

**Review by jaseemjaskp — COMMENTED** [review]
Automated PR review (PR Review Toolkit: code-reviewer, silent-failure-hunter, type-design-analyzer, pr-test-analyzer, comment-analyzer, code-simplifier). The core fix is correct and well-scoped; these are refinements. The non-string-key `AttributeError` case is already covered by CodeRabbit's comment and is not repeated here.

File: unstract/core/src/unstract/core/plugins/plugin_manager.py [hunk 6]
```diff
@@ -215,13 +215,13 @@ def load_plugins(self) -> None:
...
+                        plugin_data[key] = value
```
[the remark this answers is not in the document]
Comment (jaseemjaskp) [reply unstract/core/src/unstract/core/plugins/plugin_manager.py]: Fixed in f0ccfc8ce — added an `isinstance(key, str)` guard before the suffix check, so a non-string metadata key is skipped instead of raising an uncaught `AttributeError` that would otherwise crash `load_plugins()` for every plugin.

File: unstract/core/src/unstract/core/plugins/plugin_manager.py (hunk 1 above)
Comment (jaseemjaskp) [reply unstract/core/src/unstract/core/plugins/plugin_manager.py:225]: Leaving this as a follow-up rather than fixing here — a real gap, but the proper fix (validating required class fields at load time, or narrowing the `except Exception` blocks in `api_hub_usage_utils.py` to distinguish a misconfigured plugin from a transient error) touches consumer files outside this PR's diff. Noted in the PR description.

File: unstract/core/src/unstract/core/plugins/plugin_manager.py (hunk 2 above)
Comment (jaseemjaskp) [reply unstract/core/src/unstract/core/plugins/plugin_manager.py]: Fixed in f0ccfc8ce — hoisted the suffix tuple to a module-level `_CLASS_HANDLE_SUFFIXES` constant and trimmed the comment so it no longer calls out `headers_cache_class` specifically.

File: unstract/core/src/unstract/core/plugins/plugin_manager.py (hunk 3 above)
Comment (jaseemjaskp) [reply unstract/core/src/unstract/core/plugins/plugin_manager.py]: Fixed in f0ccfc8ce — added a one-line comment noting `key not in plugin_data` is defensive/future-proofing and not currently reachable.

File: unstract/core/tests/test_plugin_manager.py (hunk 4 above)
Comment (jaseemjaskp) [reply unstract/core/tests/test_plugin_manager.py:56]: Added `test_class_field_propagates_regardless_of_value` in f0ccfc8ce, which asserts `headers_cache_class: None` still gets propagated onto the plugin dict.

File: unstract/core/tests/test_plugin_manager.py (hunk 5 above)
Comment (jaseemjaskp) [reply unstract/core/tests/test_plugin_manager.py:95]: Added `test_inactive_plugin_is_not_loaded` in f0ccfc8ce, mirroring the existing `disable=True` test for the `is_active=False` branch.

**Review by chandrasekharan-zipstack — APPROVED** [review]

**Review by athul-rs — APPROVED** [review]
